rust/src/dataset/cleanup.rs

```diff
@@ -0,0 +1,697 @@
+                }
+            }
+            Some("arrow") => {
+                if relative_path.as_ref().starts_with("_deletions") {
+                    files_list.delete_file_paths.push(relative_path);
+                }
+            }
+            Some("bin") => {
+                if relative_path.as_ref().starts_with("_deletions") {
+                    files_list.delete_file_paths.push(relative_path);
+                }
+            }
+            _ => (),
+        };
+        Ok(())
+    }
+}
+
+pub async fn cleanup_old_versions(dataset: &Dataset, before: DateTime<Utc>) -> Result<()> {
+    let mut cleanup = CleanupTask::new(dataset, before);
+    cleanup.run().await
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::{Arc, Mutex};
+
+    use arrow_array::RecordBatchReader;
+    use chrono::Duration;
+
+    use crate::{
+        dataset::{ReadParams, WriteMode, WriteParams},
+        index::{
+            vector::{MetricType, StageParams, VectorIndexParams},
+            DatasetIndexExt, IndexType,
+        },
+        io::{
+            object_store::{ObjectStoreParams, WrappingObjectStore},
+            ObjectStore,
+        },
+        utils::{
+            datagen::{some_batch, BatchGenerator, IncrementingInt32},
+            temporal::utc_now,
+            testing::{assert_err_containing, MockClock, ProxyObjectStore, ProxyObjectStorePolicy},
+        },
+        Error, Result,
+    };
+    use all_asserts::{assert_gt, assert_lt};
+    use tempfile::{tempdir, TempDir};
+
+    use super::*;
+
+    #[derive(Debug)]
+    struct MockObjectStore {
+        policy: Arc<Mutex<ProxyObjectStorePolicy>>,
+    }
+
+    impl WrappingObjectStore for MockObjectStore {
+        fn wrap(
+            &self,
+            original: Arc<dyn object_store::ObjectStore>,
+        ) -> Arc<dyn object_store::ObjectStore> {
+            Arc::new(ProxyObjectStore::new(original, self.policy.clone()))
+        }
+    }
+
+    impl MockObjectStore {
+        pub(crate) fn new() -> Self {
+            Self {
+                policy: Arc::new(Mutex::new(ProxyObjectStorePolicy::new())),
+            }
+        }
+    }
+
+    #[derive(Debug, PartialEq)]
+    struct FileCounts {
+        num_data_files: usize,
+        num_manifest_files: usize,
+        num_index_files: usize,
+        num_delete_files: usize,
+    }
+
+    struct MockDatasetFixture<'a> {
+        // This is a temporary directory that will be deleted when the fixture
+        // is dropped
+        _tmpdir: TempDir,
+        dataset_path: String,
+        mock_store: Arc<MockObjectStore>,
+        pub clock: MockClock<'a>,
+    }
+
+    impl<'a> MockDatasetFixture<'a> {
+        fn try_new() -> Result<Self> {
+            let tmpdir = tempdir()?;
+            // let tmpdir_uri = to_obj_store_uri(tmpdir.path())?;
+            let tmpdir_path = tmpdir.path().as_os_str().to_str().unwrap().to_owned();
+            println!("A={tmpdir:?}");
+            println!("B={tmpdir_path:?}");
```

Comment:
Consider using the debug macro, which will print the name, debug value, and the line number where this print came from.
```suggestion
            dbg!(&tmpdir);
            dbg!(&tmpdir_path);
```

Reply:
Well, these prints shouldn't even be in here :frowning:.  They've been removed.

But yes, the `dbg` macro is good to know..